Restore struct info in PDF output

annda opened the pull request:
fixes #300
related to #424 

It adds `export_pdf` to whitelisted actions in struct's render method. Is this the right approach or a band-aid solution (see #424)?

splitbrain commented:
Hmm I think we could make this a little bit more robust (in terms of what people expect) by instead of checking for the current action checking if the renderer is a descendant of `Doku_Renderer_xhtml`.

Or maybe we could use a blacklist instead of a whitelist, since all we don't want to run on is `metadata`, right?

@micgro42, @solewniczak do you have any opinion on this?

micgro42 commented:
It has been some time since I've looked into this, but aren't there some more actions around? I'm thinking of raw text or odt export. Also, I'm not sure I remember how exactly we handled this in the prosemirror plugin?

I think I would favor opening this gradually. Maybe create an explicit `$whitelistedActions` variable? Then we can expand it more easily when requested by other plugins.
On the other hand doing the check on `Doku_Renderer_xhtml` shouldn't break too much either, I hope.

solewniczak commented:
Checking if the renderer is a descendant of `Doku_Renderer_xhtml ` will be most robust in my opinion.

annda commented:
I changed the PR to check the renderer type (`Doku_Renderer_xhtml`). I also assumed that all the XTML renderers keep the `xhtml` format/mode, so I removed the mode checks as well.

But now I am no longer sure if it is the best way to go:

Most of the renderer plugins I checked do inherit from `Doku_Renderer_xhtml`. Exceptions like odt or latexport do not work with the present action whitelist either.

It is an improvement over the current state, but perhaps we need another solution?

annda commented:
Changed whitelist to blacklist after a quick discussion of https://github.com/cosmocode/dokuwiki-plugin-struct/pull/446#issuecomment-483618116 with @splitbrain

solewniczak commented:
This merge causes that struct data box is shown on actions such as "login", "admin" etc. I think than rather  than using "BLACKLIST_RENDERER" we should use "BLACKLIST_ACT" constant and filter out the actions where struct data box shouldn't be visible.
![struct-data-box-on-login-page](https://user-images.githubusercontent.com/468845/57908537-82603180-7880-11e9-8d31-381a2e21ad57.png)

splitbrain commented:
Yeah, I totally forgot about texts rendered during the same request that are unrelated to the main page. I'll open a new PR.
